### PR TITLE
Ben's Review

### PR DESCRIPTION
**Pull request by benhoar**

### CODE REVIEW

File: Train.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+#Parameterized function for training models, returns accuracy
+
+# BEN Train and test -> convert to percentage of available data
+def findAccuracy(search_space):
+    lr,wd=search_space["learning"],search_space["weight_decay"]
+    model = twoDRes(lr,wd)
+    train, test, _, _ = load_data(
+            train_batch_size=256, test_batch_size=4096, train_size=2300, 
+            n_crv=6, noise_mag=0.15, noise_mag_train=0.15, mask_sr=False, test_size=1000, inp=inp, peaknoise_mag=0)
```
Comment (benhoar) [inline Train.py:31]: I would suggest converting all train/test size values to percentage based values. You could add a method to calculate the values in these lines and pass them as parameters.

File: Train.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    loss,test_loss = model.fit(train, test, epoch=100)
+    target, pred = model.test(test, epoch=1)
+    return accuracy_score(target, pred)
+
+if __name__ == '__main__':
+    #input prompt
+
+    # BEN clearer directions on how to select models
+    print("What mechanisms to classify:")
+    inp=input("E "+"EC1 "+"CE "+"DL "+"SR "+"T "+"ECP "+"DISP "+"ECE \n")
```
Comment (benhoar) [inline Train.py:41]: Add explicit directions on how users are to enter mechanisms, add checking to ensure valid input, and allow for users to retry in case there is a mistake in their input. I would also parse their and standardize the format of model names now, e.g. sort the user inputs alphabetically so that "DL, E" is equivalent to "E, DL." Further, allow users to input a degree of training noise (we should also allow for non-uniform noise as in some training data samples from a noise distribution with low standard deviation, others with high). Finally, ask users to provide the path to their data folders – advise them to ONLY have data in the root folder of their data directory.

File: Train.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+    #check whether model is already saved in directory
+    # BEN use clearer names {dirlist: done_models, i:, model, }
+    dirlist = os.listdir('./trainedModels')
+    flag=0
+    for i in dirlist:
+        if i==inp:
+                flag=1
```
Comment (benhoar) [inline Train.py:49]: os actually has a useful method here – look into os.path and specifically here os.path.isfile

File: Train.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    
+    #load data and initialize model
+    train, test, _, _ = load_data(
+            train_batch_size=256, test_batch_size=4096, train_size=2300, 
+            n_crv=6, noise_mag=0.15, noise_mag_train=0.15, mask_sr=False, test_size=1000, inp=inp, peaknoise_mag=0)
+    model = twoDRes(1e-3,1e-2)
+    os.chdir('./trainedModels')
```
Comment (benhoar) [inline Train.py:56]: I won't tell you what to do here, but be sure you want to/need to change directories at runtime, if I were you I would be concerned about the stability of relative path names. For example, say I want to save models in root/models but I have changed to a new directory at some point and my current directory has no folder root/models.

File: Train.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+    #if model is new, train one, otherwise return stored model
+    if flag==0:
+        os.mkdir(inp)
+        os.chdir('./'+inp)
```
Comment (benhoar) [inline Train.py:61]: kind of preference but it does sometimes make things easier, the current pythonic way to make composite strings is with f strings, for example, x = 5, I can make a string using x with the format f'the value for x is {x}'

File: Train.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        loss, test_loss = model.fit(train, test, epoch=100)
+        torch.save(model.model.state_dict(), inp)
+        target, pred = model.test(test, epoch=1)
+        f=open("Confusion_Matrix", "x")
+        f.write(str(confusion_matrix(target,pred)))
+        f.close()
```
Comment (benhoar) [inline Train.py:67]: It is typically ill-advised in python to use files in this way – instead, the "correct" way is to use 
`with open(myfile, 'w') as filevariable:
        filevariable.write("hey there")`
The reason is that if there is an error, this file will close using this syntax, whereas in the syntax you have, an error writing to file will not close the file

File: Train.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+        for i in accuracy:
+            f.write(str(i))
+            f.write("\n")
+        f.close()
+
+    #hyperparameter optimization space
+    space = {
+        "learning": hp.lognormal("learning",-3,1),
+        "weight_decay": hp.lognormal("weight_decay",-2,1),
+    }
```
Comment (benhoar) [inline Train.py:92]: Are these the only two variables worth optimizing? They may be, I don't know, just highlighting in case you want to add more.

File: data_processing.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+import random
+
+length = 1000
+path = './Data2/'
+folder = {
+    'CE': path+'CE', 'EC1': path+'EC1', 'E': path+'E',
+    'ECE': path+'ECE', 'DISP': path+'DISP', 'DL': path+'DL','ECP': path+'ECP', 'T': path+'T'
+    #'CE': path+'CE/', 'EC': path+'EC/', 'E': path+'E/',
+    #'ECE': path+'ECE/', 'DISP': path+'DISP/'
+}
```
Comment (benhoar) [inline data_processing.py:18]: Related to user input – we should have a means of allowing users to provide multiple jobs at once. For example, if they provide JUST the root directory for the data, but then tell us which mechanisms to use, then that is easier for everyone because we don't have to hard code in folders and they don't have to duplicate data and make many data folders etc. I would look into os.scandir – this allows for really nice directory traversal, given the root directories and the names of data folders to use for an experiment, we could have a really convenient user interface.

Ex. root dir  = 'users/bhoar/mydata/', mechs = ['E', 'EC']
We can use os.scandir to look in that dir for folders named 'E' and 'EC'

File: data_processing.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+            #mat = np.stack((mat_val1, mat_val2), axis=1)
+            dataset.append({'data': 
+                mat, 'label': e, 'file': f, 'key': j})
+        print(scanlist)
+        with open(path+'{}.pkl'.format(e), 'wb') as f:
+            pickle.dump(dataset, f)
+
+
+class Data(Dataset):
+    def __init__(self, array, train=False, n_crv=6, noise_mag=1, mask_sr=False, peaknoise_mag=0):
```
Comment (benhoar) [inline data_processing.py:61]: Weitong's mask_sr parameter here can be improved. We can make this more flexible by making it a user requested parameter (log, lin, or mixed scan rates).

File: data_processing.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+        # dataformat: 6 * n * 1000
+        self.data = array
+        self.train = train
+        self.n_crv = n_crv
```
Comment (benhoar) [inline data_processing.py:65]: the n_crv variable is not used in the code because it is always 6. Up to you if you want to account for future scenarios where 6 is not required, but as it stands we always demand 6.

File: data_processing.py [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+        """ mask = torch.zeros_like(entry)
+        if ncrv == 1:
+            mask[np.random.randint(6)] = 1
+        else:
+            mask[0] = 1
+            mask[5] = 1
+            arr = np.random.choice([1, 2, 3, 4], size=(ncrv - 2), replace=False)
+            mask[arr] = 1
+        if self.mask_sr:
+            mask[:, 2, :] = 0 """
```
Comment (benhoar) [inline data_processing.py:99]: ok to delete this if you want

File: data_processing.py [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+        if self.train:
+            return {'data': ((noise + peaknoise + entry) * mask), 'label': label}
+        else:
+            return {'data': ((noise + peaknoise + entry) * mask), 'label': label}
```
Comment (benhoar) [inline data_processing.py:103]: redundant

File: Train.py [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+Created on Sat Aug 20 15:51:06 2022
+
+@author: Pranav
+"""
+
+from hyperopt import tpe, hp, fmin, STATUS_OK,Trials
+from hyperopt.pyll.base import scope
+import os
+from data_processing import load_data, folder
+from network import Resnet50 as threeDRes
```
Comment (benhoar) [inline Train.py:12]: I think we are using only twoDRes, in general there is a lot of weird stuff from Weitong in terms of naming and imports and stuff, up to you whatever changes you want to make – trust yourself if you see something weird and change it if you want

**Review by benhoar — COMMENTED** [review]
Good job! In general, the biggest things you need to consider are:

1. What should user input look like and what parameters can they tune (classes, noise, type of scan rate sampling used, etc)?
2. What should you save during training?
- User inputs
- Training loss versus epoch
- Bayesian optimization tuning information
- Model accuracy, confusion matrix, etc., lots of room for creativity here to save yourself time later, data files related to training are cheap to make and store and expensive to backtrack and obtain, the more you save the better (especially during development)
- In my opinion, verbose file names are not always the worst thing, I would rather have a file named resnet_layers_50_classes_E_EC_CE_accuracy_98_noise_10_20230201 than model_E_EC_CE
